### ISSUE TITLE
Knora IRI's (URI - URL - URN) some thoughts in preparation of next weeks meeting

### ISSUE THREAD

**Issue by subotic** [dhlab-basel/Knora#626]
Following up on @benjamingeer post in #610:

> I think we should talk about IRI formats again with Lukas. The last discussion we had is in #595,
> where we basically decided:
>
> - Only resource IRIs need to be dereferencable.
> - Resource IRIs will be handle.net IRIs.
> - Therefore there's no need for rdfh.ch.
> - Since other IRIs don't need to be dereferencable, they can be whatever we want. We need to decide what they should be.

**Comment by subotic** [comment]
The following came to my mind: Why not use **internally** URNs for everything?

A URN would look something like this: `urn:knora:resource:[UUID]`

The reasoning is as follows. Knora is an application which will be used by the DaSCH but potentially also by others.

When we make decisions like what will be dereferenceable and what will not, then we decide this as the DaSCH. What if someone using Knora would like to have everything or nothing dereferenceable?

The solution to this problem would be to make it configurable. Since we don't allow direct access to the data stored in the triplestore, we could transform the URNs to URLs based on some configuration inside the `application.conf` on every API call.

This would also allow us (the DaSCH) to easily change our minds in the future, about what should and should not be dereferenceable.

**Comment by benjamingeer** [comment]
@subotic I like that idea, because I'm actually a bit worried about using `handle.net` IRIs internally. What if `handle.net` disappears?

However, if we have different internal and external IRIs, wouldn't it be easier just to keep using `http://rdfh.ch` internally? That way, people wouldn't have to change their existing data to use URNs.

Also, we still have to answer the question: what IRI should the API return for something that's not dereferenceable?

**Comment by loicjaouen** [comment]
What is the definition of _internally_ here? internally as in within the scala code and triplestore and not dereferenceable?

**Comment by benjamingeer** [comment]
Internally means in the triplestore, externally means what we return in the API.

Our last idea was to use dereferenceable IRIs both internally and externally. Ivan is suggesting that we could have non-dereferenceable resource IRIs internally, and convert them on the fly to dereferenceable ones as needed in the API.

This would be similar to what we're already doing for resource class and property IRIs in version 2 of the API.

**Comment by mrivoal** [comment]
Yes, but what about the idea of a simplified public sparql endpoint for Knora? Honestly, I have no idea as to how it should be done, but if we keep this idea, it may not be enough to have dereferencable IRIs served by the API. A Sparql endpoint should probably provide dereferencable IRIs, right?

**Comment by subotic** [comment]
> Yes, but what about the idea of a simplified public sparql endpoint for Knora?

We would need to create a second triplestore with cleaned up (removed permissions, etc.) data anyway. In the same step, we could then transform every URN into a URL.

**Comment by subotic** [comment]
> However, if we have different internal and external IRIs, wouldn't it be easier just to keep using http://rdfh.ch internally? That way, people wouldn't have to change their existing data to use URNs.

The only reason to change internal IRIs to URNs would be to make it clearer that we are actually looking at an identifier which is not dereferenceable. For example,  having something like `urn:knora:users:007` versus `http://rdfh.ch/users/007`.

As far as the rewriting on the fly goes, both URLs and URNs will work the same.

**Comment by benjamingeer** [comment]
> The only reason to change internal IRIs to URNs would be to make it clearer that we are actually looking at an identifier which is not dereferenceable.

That sounds good to me, but:

* Let's make sure triplestores actually accept URNs as IRIs.
* And for resources, let's include the project ID as we do now, to reduce the risk of UUID collisions.

**Comment by loicjaouen** [comment]
@subotic : it makes sense (it means maintenance for us to move the data though)

**Comment by subotic** [comment]
> Let's make sure triplestores actually accept URNs as IRIs.

Data:

```
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
INSERT DATA {
	GRAPH <example:ns:graph> {
	    <example:ns:A> rdf:type owl:Class .
            <example:data:1> rdf:type <example:ns:A> .
	}
}
```

Query:

```
select *
from <example:ns:graph>
where { 
    ?s ?p ?o .
} limit 100 
```

Works both on `GraphDB` and `AllegroGraph`.

**Comment by benjamingeer** [comment]
cool!

**Comment by loicjaouen** [comment]
We discussed this issue with @mrivoal , @benjamingeer and @lrosenth and came to the conclusion that we should keep [IRI](https://www.w3.org/TR/rdf11-primer/#section-IRI).

We liked the idea of URN to make sure that they are not dereferenceable and clearly set the limit of Knora's inside and outside world.

But we decided to stick to standards as much as possible to limit the risk of incompatibilities with triplestores (existing triplestores or their evolutions) and ease the learning curve of Knora.
